From 37f4e1d0edbf341585f744c6cabd5e1e60b1600f Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:03:07 -0400
Subject: [usb] wusbcore: send keepalives to unauthenticated devices

Message-id: <1410887463-20674-116-git-send-email-dzickus@redhat.com>
Patchwork-id: 95399
O-Subject: [RHEL7 PATCH 115/591] usb: wusbcore: send keepalives to unauthenticated devices
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit f4042c068ae53625e4a9f39543c2693d07a4e14c
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Mon Dec 2 15:39:43 2013 -0600

    usb: wusbcore: send keepalives to unauthenticated devices

    This patch modifies the WUSB device disconnect timer code to send
    keepalives to all connected devices even if they are not authenticated.
    This fixes a problem where unauthenticated devices that lose their
    connection before they are authenticated will stay in the device tree
    forever.  More importantly, devices in this situation will never
    relinquish their port on the root hub so eventually all root ports will
    be taken up and no other devices can connect.

    A comment in the existing code states that there are some devices that
    may not respond to keepalives if they have not been authenticated.  That
    comment is about 5 years old and I don't know of any WUSB devices that
    act that way.  Either way, any buggy devices that may still be around
    will continue to work as long as they can transition to the
    authenticated state within the WUSB LOA timeout of 4s, which is not
    unreasonable to expect.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/wusbcore/devconnect.c b/drivers/usb/wusbcore/devconnect.c
index 33a1278..723d6b2 100644
--- a/drivers/usb/wusbcore/devconnect.c
+++ b/drivers/usb/wusbcore/devconnect.c
@@ -411,9 +411,6 @@ static void __wusbhc_dev_disconnect(struct wusbhc *wusbhc,
 /*
  * Refresh the list of keep alives to emit in the MMC
  *
- * Some devices don't respond to keep alives unless they've been
- * authenticated, so skip unauthenticated devices.
- *
  * We only publish the first four devices that have a coming timeout
  * condition. Then when we are done processing those, we go for the
  * next ones. We ignore the ones that have timed out already (they'll
@@ -448,7 +445,7 @@ static void __wusbhc_keep_alive(struct wusbhc *wusbhc)
 
   if (wusb_dev == NULL)
    continue;
-  if (wusb_dev->usb_dev == NULL || !wusb_dev->usb_dev->authenticated)
+  if (wusb_dev->usb_dev == NULL)
    continue;
 
   if (time_after(jiffies, wusb_dev->entry_ts + tt)) {
-- 
1.7.1